packages/eslint-plugin/tools/validate-docs/check-for-rule-docs.ts

```diff
@@ -2,23 +2,48 @@ import { TSESLint } from '@typescript-eslint/experimental-utils';
 import fs from 'fs';
 import path from 'path';
 import { logRule } from '../log';
+import chalk from 'chalk';
 
 function checkForRuleDocs(
   rules: Record<string, Readonly<TSESLint.RuleModule<string, unknown[]>>>,
 ): boolean {
-  const ruleDocs = new Set(
-    fs.readdirSync(path.resolve(__dirname, '../../docs/rules')),
-  );
+  const docsRoot = path.resolve(__dirname, '../../docs/rules');
+  const ruleDocs = new Set(fs.readdirSync(docsRoot));
 
   let hasErrors = false;
   Object.keys(rules).forEach(ruleName => {
+    const rule = rules[ruleName].meta;
+    const errors: string[] = [];
     const ruleHasDoc = ruleDocs.has(`${ruleName}.md`);
-    hasErrors = hasErrors || !ruleHasDoc;
-    logRule(
-      ruleHasDoc,
-      ruleName,
-      `Couldn't find file docs/rules/${ruleName}.md`,
-    );
+    if (!ruleHasDoc) {
+      errors.push(`Couldn't find file docs/rules/${ruleName}.md`);
+    } else {
+      const file = fs.readFileSync(
+        path.join(docsRoot, `${ruleName}.md`),
+        'utf-8',
+      );
+      const match = /^\s*#[^\r\n]*/.exec(file) ?? [];
+      const expectedDescription = `# ${rule.docs.description} (\`${ruleName}\`)`;
+      if (!match[0] || match[0] !== expectedDescription) {
+        errors.push(
+          'Header of file does not match the rule metadata.',
+          `    Expected: ${chalk.underline(expectedDescription)}`,
+          `    Received: ${chalk.underline(match[0])}`,
+        );
+      }
```

Comment:
i would probably prefer splitting the `!match[0]` case out, and explicitly state "No header of file", so things are a bit clearer (not that this should ever happen)